reproman/distributions/tests/test_singularity.py

```diff
@@ -7,56 +7,43 @@
 #
 # ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
 
-import os
-import tempfile
+import os.path as op
+
+import pytest
 
 from ...cmd import Runner
-# from ...distributions.singularity import SingularityDistribution
-# from ...distributions.singularity import SingularityImage
 from ...distributions.singularity import SingularityTracer
+from ...support.external_versions import external_versions
 from ...tests.skip import mark
 
 
 @mark.skipif_no_network
 @mark.skipif_no_singularity
-def test_singularity_trace():
-
+@pytest.mark.xfail(
+    external_versions["cmd:singularity"] >= '3',
+    reason="Pulling with @hash fails with Singularity v3 (gh-406)")
+def test_singularity_trace(tmpdir):
+    tmpdir = str(tmpdir)
     # Download and set up singularity image file
-    runner = Runner()
-    runner.run(['singularity', 'pull',
-        'shub://vsoch/hello-world@42e1f04ed80217895f8c960bdde6bef4d34fab59'])
-    image_file = tempfile.mktemp()
-    os.rename('vsoch-hello-world-master-latest.simg', image_file)
+    runner = Runner(cwd=tmpdir)
+    expected_md5sum = "96563a1ee2a3d9dbd082efb8d263fc09"
+    location = "singularityhub/busybox@{}" + expected_md5sum
+    runner.run(['singularity', 'pull', '--name', 'img', 'shub://' + location])
```

Comment:
> It looks like the pulling by hash fail for versions > 3 is limited to shub and not docker hub: [...]

Interesting.

> We could pull an image to use in testing from docker hub rather than shub:

I'd vote to stick to an shub url for a few reasons:

  * Pulling from docker:// doesn't contain much useful metadata for us to inspect (e.g., there doesn't seem to be a size or base image, and we can't expect a certain build date because the build happens at pull).

  * The docker:// pull takes more time because the build happens at pull.

  * I think there's value in marking something that should work as broken.  And there's an unmerged upstream fix (3652), so it probably won't be broken indefinitely.